_posts/2021-09-26-where-the-store-is-stored.md

```diff
@@ -0,0 +1,292 @@
+somewhere in our app's entrypoint and pass it along everywhere.
+
+```js
+// index.js
+
+function startApp(store) {
+  store.add('new car');
+
+  document.body.innerText = store.list().join(', ');
+
+  // ...
+}
+
+function initApp() {
+  const store = buildStore();
+
+  startApp(store);
+}
+
+initApp();
+```
+
+Of course, any function that needs to access the store will have to receive it
+via its parameters... We're right back where we left our props drilling issue.
+
+# Pros and Cons of each store location 
+
+OK so we've outlined three potential store locations. You can see me coming with
+the list of pros and cons for each of them. 
+
+Let's start with modules. Pros: it's raw ES6 (`import` / `export`) and it is
+quite simple to grasp. It's also very flexible. We can do more or less whatever
+we want in a module and every part of our codebase can use it.
+
+But there is a big constraint. Not all codebases always run in the same runtime.
+Actually a lot of single page applications run in both browser and NodeJS
+runtimes, where the contexts are very different. In the browser, we can make the
+assumption that there is only one user at a given time using the runtime. So our
+module state will be "our user state". On NodeJS, we will probably use the code
+to render the page server-side for multiple HTTP requests. In this situation,
+our module state will be shared between multiple. Boom! Enjoy the data leak
+security issue.
+
+For library or framework locations such as our React `useState`, the big
+advantage is probably the strong integration with the corresponding library.
+Even bigger store solutions like Redux come with their binding like
+`react-redux` that does the job for us. No need to handle the Context / Provider
+stuff I mentioned. No issue with server-side rendering: the store being "in
+React runtime", it will be scoped to the `ReactDOM.render` or
+`ReactDOMServer.renderToString` (which will be called once for every http
+requests).
+
+I'm beginning to think this deep integration with view librairies in a way that
+makes server and browser contexts similar is a big factor in the success of
+Redux and others (Mobx, etc.).
+
+The main issue is however that we're locked to the corresponding library /
+framework.  Let's say you have an SPA where you want React to really only be
+about the views, and that you need to read / update / subscribe to store change
+in other contexts.  You just can't, it's React or bust. Back at square one.
+
+Let's give another chance to our programmatic store sharing solution, ie.
+"crafting a store as soon as the app starts and then pass it along to every
+subsequent bit of code that needs it through a sort of a handmade dependency
+injection".
+
+It's clearly not very convenient to have to pass our store programaticaly to
+every part of the codebase as it creates a lot of boilerplate, but at least
+we face no issue with SSR if we craft one store per request server-side, and one
```

Comment:
It doesn't cost much to write Server Side Rendering here imo.

Reply:
Yes that's not really the point of this paragraph: "we face no issue with SSR". Should I rephrase it?